docs/source/caching/index.rst

```diff
@@ -0,0 +1,115 @@
+
+Caching is also implemented for Data nodes. This is not very useful in practice (yet), but is an easy way to show how the caching mechanism works:
+
+.. ipython::
+    :verbatim:
+
+    In [1]: from __future__ import print_function
+
+    In [2]: from aiida.orm.data.base import Str
+
+    In [3]: n1 = Str('test string')
+
+    In [4]: n1.store()
+    Out[4]: u'test string'
+
+    In [5]: n2 = Str('test string')
+
+    In [6]: n2.store(use_cache=True)
+    Out[6]: u'test string'
+
+    In [7]: print('UUID of n1:', n1.uuid)
+    UUID of n1: 956109e1-4382-4240-a711-2a4f3b522122
+
+    In [8]: print('n2 is cached from:', n2.get_extra('cached_from'))
+    n2 is cached from: 956109e1-4382-4240-a711-2a4f3b522122
+
+As you can see, passing ``use_cache=True`` to the ``store`` method enables using the cache. The fact that ``n2`` was created from ``n1`` is stored in the ``cached_from`` extra of ``n2``.
+
+When running a ``JobCalculation`` through the ``Process`` interface, you cannot directly set the ``use_cache`` flag when the calculation node is stored internally. Instead, you can pass the ``_use_cache`` flag to the ``run`` or ``submit`` method.
+
+Caching is **not** implemented for workchains and workfunctions. Unlike calculations, they can not only create new data nodes, but also return exsting ones. When copying a cached workchain, it's not clear which node should be returned without actually running the workchain. This is explained in more detail in the section :ref:`caching_provenance`.
+
+Configuration
+-------------
+
+Of course, using caching would be quite tedious if you had to set ``use_cache`` manually everywhere. To fix this, the default for ``use_cache`` can be set in the ``.aiida/cache_config.yml`` file. You can specify a global default, or enable / disable caching for specific calculation or data classes. An example configuration file might look like this:
+
+.. code:: yaml
+
+    profile-name:
+      default: False
+      enabled:
+        - aiida.orm.calculation.job.simpleplugins.templatereplacer.TemplatereplacerCalculation
+        - aiida.orm.data.base.Str
+      disabled:
+        - aiida.orm.data.base.Float
+
+This means that caching is enabled for ``TemplatereplacerCalculation`` and ``Str``, and disabled for all other classes. In this example, manually disabling ``aiida.orm.data.base.Float`` is actually not needed, since the default value for all classes is already ``False``. Note also that the fully qualified class import name (e.g., ``aiida.orm.data.base.Str``) must be given, not just the class name (``Str``). This is to avoid accidentally matching classes with the same name.
```

Comment:
The following line: "since the default value for all classes is already ``False``" may be a bit confusing. By default all `Node` classes are in fact marked as cacheable, in terms of implementation. Only `Calculation` nodes that are not `InlineCalculation` or `JobCalculation` have `_cacheable = False`. But in this line I think you refer to the fact that in this particular config example, the default for caching globally is set to `False`. However, with this explanation a user might think that if he sets the default to `True` in the config, caching is still going to be disabled for the `Float` class, when in reality it is not.

Maybe we can write that in general all node class are cacheable, except if the default in the config is set to `False` or the class is specifically disabled. With the additional exception that `WorkCalculations` and their derivatives are always disabled and cannot be enabled through the configuration.

Reply:
I changed the wording a bit to hopefully make it clear that it's referring to the default configuration value. As for the ``_cacheable`` property, I don't think we need to tell the user about this at that point. The fact that workchains and workfunctions cannot be cached is mentioned in the introductory section.

The user should need to know about ``_cacheable`` only if he wants to disable it manually for his classes, which is what the "how are cached nodes matched" section is for.